evap/results/tools.py

```diff
@@ -171,42 +170,101 @@ def _calculate_results_impl(course):
+    if not weights:
+        return normalize_distribution(sum_distributions(distributions))
 
-    return final_avg, final_dev
+    weighted_distributions = []
+    for distribution, weight in zip(distributions, weights):
+        weighted_distributions.append(multiply_distribution(distribution, weight))
+    return normalize_distribution(sum_distributions(weighted_distributions))
+
+
+def calculate_average_grade_and_distribution(course):
+    # will contain a list of results for each contributor and one for the course (where contributor is None)
+    grouped_results = defaultdict(list)
+    for __, contributor, __, results, __ in calculate_results(course):
+        grouped_results[contributor].extend(results)
+
+    # early return if no rating results exist for this course
+    if not any(result.average for result in grouped_results[contributor] for contributor in grouped_results.keys() if result.question.is_rating_question):
+        return None, None
+
+    contributor_averages = []
+    contributor_distributions = []
+    course_grade_questions_average = None
+    course_grade_questions_distribution = None
+    course_non_grade_questions_average = None
+    course_non_grade_questions_distribution = None
+
+    for contributor, results in grouped_results.items():
+        grade_question_results = [result for result in results if result.question.is_grade_question]
+        grade_questions_average = avg([result.average for result in grade_question_results])
+        grade_questions_distribution = avg_distribution([result.counts for result in grade_question_results])
+        non_grade_question_results = [result for result in results if result.question.is_non_grade_rating_question]
+        non_grade_questions_average = avg([result.average for result in non_grade_question_results])
+        non_grade_questions_distribution = avg_distribution([result.counts for result in non_grade_question_results])
+        if contributor:
+            contributor_averages.append(avg(
+                [grade_questions_average, non_grade_questions_average],
+                [settings.CONTRIBUTOR_GRADE_QUESTIONS_WEIGHT, settings.CONTRIBUTOR_NON_GRADE_QUESTIONS_WEIGHT]
+            ))
+            contributor_distributions.append(avg_distribution(
+                [grade_questions_distribution, non_grade_questions_distribution],
+                [settings.CONTRIBUTOR_GRADE_QUESTIONS_WEIGHT, settings.CONTRIBUTOR_NON_GRADE_QUESTIONS_WEIGHT]
+            ))
+        else:
+            course_grade_questions_average = grade_questions_average
+            course_grade_questions_distribution = grade_questions_distribution
+            course_non_grade_questions_average = non_grade_questions_average
+            course_non_grade_questions_distribution = non_grade_questions_distribution
+
+    contributors_average_grade = avg(contributor_averages)
+    contributors_average_distribution = avg_distribution(contributor_distributions)
```

Comment:
maybe `contributors_averages` for the other one, but i wouldn't call this `contributor_average_grade`  because its the average grade of all contributors